modules/siddhi-core/src/main/java/org/wso2/siddhi/core/query/processor/stream/window/SessionWindowProcessor.java

```diff
@@ -0,0 +1,549 @@
+import org.wso2.siddhi.query.api.definition.Attribute;
+import org.wso2.siddhi.query.api.exception.SiddhiAppValidationException;
+import org.wso2.siddhi.query.api.expression.Expression;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Implementation of {@link WindowProcessor} which represent a Window operating based on a session.
+ */
+@Extension(
+        name = "session",
+        namespace = "",
+        description = "A session window that holds events into a session "
+                + "which has a grouping attribute (session key). After a session gap period, "
+                + "the session window would be expired. When a new event "
+                + "arrives with a value for session key, then it matches with a session window "
+                + "which has the same session key. A latency can be added to late arrival events "
+                + "to merge with the previous session windows. That latency time period "
+                + "should be less than the session gap time period. To have aggregate functions with session window, "
+                + "a 'group by' should be done with the session key. ",
+        parameters = {
+                @Parameter(name = "window.session",
+                        description = "The session gap time period(sec, min, ms).",
+                        type = {DataType.INT, DataType.LONG, DataType.TIME}),
+                @Parameter(name = "window.key",
+                        description = "The grouping attribute for events.",
+                        type = {DataType.STRING}, optional = true, defaultValue = "default-key"),
+                @Parameter(name = "window.allowedlatency",
+                        description = "This will allow to alive session window for a defined time period",
+                        type = {DataType.INT, DataType.LONG, DataType.TIME}, optional = true, defaultValue = "0")
+        },
+        examples = {
+                @Example(
+                        syntax = "define stream purchaseEventStream "
+                                + "(user string, item_number int, price float, quantity int);\n"
+                                + "\n"
+                                + "@info(name='query0) \n"
+                                + "from purchaseEventStream#window.session(5 sec, user, 2 sec) \n"
+                                + "select * \n"
+                                + "insert all events into outputStream;",
+                        description = "This will processing events that arrived from purchaseEvent stream "
+                                + "with the user name as the session key and the session gap 5 seconds. "
+                                + "It keeps the session window for 2 sec after session timeout."
+                )
+        }
+)
```

Comment:
Can you put java doc method comments?

Reply:
Thanks for the review. Done according to the comment.